### PR TITLE
Wa-219 complete implementation of point cloud data over wifi

### PR DESCRIPTION
**Pull request by jkey-Ainstein**
# Summary
Bringing in Avik's TLV format updates & completing support of Point Cloud over Wifi with the API.

Note: Point Cloud output is a development mode feature at this time. It requires restarting the WAYV Air to exit point cloud mode.

# Detailed Description of Code Changes
## Finish updating PCL over Wifi
- Brought in Avik's changes for TLV Updates
- Corrected the workMode command to match 3.0.2 FW
- Corrected some typos / bugs from the copyleft updates
- Updated the config to match something more recent

# Test Updates
## Unit Tests
N/A

## Integration Tests
N/A

# Script Updates
## Build Scripts
N/A

## Test Scripts
N/A

## Continuous Integration Scripts
N/A

# Documentation Updates
N/A

# Testing
## Development Testing
Performed testing using hardware on desk. Was able to consistently see point cloud data returned.

Note: During testing I did run into an issue where `staticRACfarCfg` didn't get written to the device the first time I ran the cfg command. Bug is created to track the issue.

### CODE REVIEW

File: worker/msg/msg_detail.py [hunk 1]
```diff
@@ -384,23 +391,27 @@ def parse_data(data):
...
+            t_dim_x             = struct.unpack('<f', data[i + 28: i + 32])[0]
+            t_dim_y             = struct.unpack('<f', data[i + 32: i + 36])[0]
+            t_dim_z             = struct.unpack('<f', data[i + 36: i + 40])[0]
+            t_ec_matrix         = list(struct.unpack('<16f', data[i + 40: i + 104]))
+            t_gating_gain       = struct.unpack('<f', data[i + 104: i + 108])[0]
+            t_confidence_level  = struct.unpack('<f', data[i + 108: i + 112])[0]
+            cp_count = 0 # Not used
+            is_target_static = 0 # Not used
+            thre = 0 # Not used
+            active2_free_count = 0 # Not used
```
Comment (AndyBorders) [inline worker/msg/msg_detail.py:411]: If the are not used why are they here?

File: worker/msg/msg_tlv.py [hunk 2]
```diff
@@ -24,102 +24,129 @@
...
+                    #     msg = MsgTlv()
+                    #     msg.header          = struct.unpack('<Q', cache[0: 8])[0]
+                    #     msg.version         = struct.unpack('<I', cache[8: 12])[0]
+                    #     msg.packet_length   = packet_len
+                    #     msg.platform        = struct.unpack('<I', cache[16: 20])[0]
+                    #     msg.frame_number    = struct.unpack('<I', cache[20: 24])[0]
+                    #     msg.timestamp       = struct.unpack('<I', cache[24: 28])[0]
+                    #     msg.numDetectedObj  = struct.unpack('<I', cache[28: 32])[0]
+                    #     msg.num_tlv         = struct.unpack('<I', cache[32: 36])[0]
+                    #     msg.sub_frame_number = struct.unpack('<I', cache[36: 40])[0]
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:78]: Do we still need the commented-out code?

File: worker/msg/msg_tlv.py [hunk 3]
```diff
@@ -24,102 +24,129 @@
...
+                        msg = MsgTlv()
+                        msg.header           = magic
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:80]: Could we use a better name than "magic" please?

File: worker/msg/msg_tlv.py [hunk 4]
```diff
@@ -24,102 +24,129 @@
...
+                        
+                        cache = cache[frameHeaderLen:]
+                        bytes_processed += frameHeaderLen
+                        
+                        for i in range(msg.num_tlv):
+                            tlv_type = struct.unpack('<I', cache[:4])[0]
+                            tlv_len = struct.unpack('<I', cache[4:8])[0]
+                            tlv_len += 8 #parse functions re-parse the header.
+                            # cache = cache[tlvHeaderLength:]
+                            # bytes_processed += tlvHeaderLength
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:98]: Remove commented out code if possible

File: worker/msg/msg_tlv.py [hunk 5]
```diff
@@ -24,102 +24,129 @@
...
+
+                            if tlv_len > msg.packet_length:
+                                break;
+                            # if cache_idx + tlv_len > totalPacketLen:
+                            #     break
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:103]: Remove commented out code if possible

File: worker/msg/msg_tlv.py [hunk 6]
```diff
@@ -24,102 +24,129 @@
...
+                                msg.tlvs.append(tlv)
+
+                            elif tlv_type == MsgTlv.TLV_TARGET_INDEX:
+                                tlv = MsgTargetIndex.parse_data(cache[:tlv_len])
+                                msg.tlvs.append(tlv)
+
+                            elif tlv_type == MsgTlv.TLV_PRESENCE_INDICATION:
+                                tlv = MsgPresenceIndication.parse_data(cache[:tlv_len])
+                                msg.tlvs.append(tlv)
+                                #print(msg.tlvs[-1].presence_detected)
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:120]: Remove commented out code if possible

File: worker/msg/msg_tlv.py [hunk 7]
```diff
@@ -24,102 +24,129 @@
...
+
+                            elif tlv_type == MsgTlv.TLV_TARGET_HEIGHT:
+                                tlv = MsgTargetHeight.parse_data(cache[:tlv_len])
+                                msg.tlvs.append(tlv)
+                                #print(vars(msg.tlvs[-1].target_heights[0]))
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:125]: Remove commented out code if possible

File: worker/msg/msg_tlv.py [hunk 8]
```diff
@@ -24,102 +24,129 @@
...
-                                print("tlv body check err")
-                                cache = cache[len(MsgTlv.HEADER_TLV):]
-                            continue
+                                print(f"Unknown TLV: {tlv_type}")
+                            cache = cache[tlv_len:]
+                            bytes_processed += tlv_len
+                        # check
+                        if bytes_processed == msg.packet_length:
+                            msgs.append(msg)
+                            # print(f"TLV Success: {tlv_type}. Total TLVs: {msg.num_tlv}")
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:133]: Remove commented out code if possible

File: worker/msg/msg_tlv.py [hunk 9]
```diff
@@ -24,102 +24,129 @@
...
                         else:
-                            break
+                            print(f"bytes_processed = {bytes_processed}, msg.pkt_length = {msg.packet_length}, msg.num_tlv = {msg.num_tlv}")
+                            print("tlv body check err")
+                            #Failed to parse TLV - Dump the next 8 bytes to reset header.
+                            # cache = cache[len(MsgTlv.HEADER_TLV):]
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:138]: Remove commented out code if possible

File: worker/msg/msg_tlv.py [hunk 10]
```diff
@@ -24,102 +24,129 @@
...
                 else:
                     break
             else:
                 # 未找到包头，不能直接清空，保留包头长度-1
-                if len(cache) >= len(MsgTlv.HEADER_TLV):
-                    cache = cache[-len(MsgTlv.HEADER_TLV) + 1: len(cache)]
+                # Failed to find magic word - Dump the next 8 bytes to reset header
+                # print(f"Couldn't find magic word in TLV. Cache Size: {cache_size}")
+                # if len(cache) >= len(MsgTlv.HEADER_TLV):
+                #     cache = cache[-len(MsgTlv.HEADER_TLV) + 1: len(cache)]
```
Comment (AndyBorders) [inline worker/msg/msg_tlv.py:149]: Remove commented out code if possible

File: worker/worker_wifi.py [hunk 11]
```diff
@@ -153,7 +156,7 @@ def handle(self):
                         self.send(bytes("getDetailTarget\n", "utf-8"))
                         ret, msg_detail = self.get_msg485(CMD_485_DETAIL_TARGET, timeout=self.comm_timeout)
                     else:
-                        self.send(bytes("getTarget\n", "utf-8"))
+                        # self.send(bytes("getTarget\n", "utf-8"))
```
Comment (AndyBorders) [inline worker/worker_wifi.py:159]: Remove commented out code if possible

File: worker/worker_wifi.py [hunk 12]
```diff
@@ -172,11 +175,32 @@ def handle(self):
...
+                            WorkerWifi.queue_msg.put((self.ip_port,msg_tlv))
+                    continue
         # 从字典删除该TCP连接
         del WorkerWifi.dict_clients[self.ip_port]
         print("Radar Disconnected:", self.ip_port)
         WorkerWifi.queue_radar_disconnect.put(self.ip_port)
 
+    def get_msgTLV(self,timeout=0.2):
+        time_now = time.time()
+        while time.time() - time_now < timeout:
```
Comment (AndyBorders) [inline worker/worker_wifi.py:193]: You should really have () and not rely on the order of operations.  
while (time.time() - time_now) < timeout:

**Review by AndyBorders — APPROVED** [review]

File: worker/msg/msg_detail.py (hunk 1 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_detail.py]: Cleaned up in latest build

File: worker/msg/msg_tlv.py (hunk 2 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py]: Removed

File: worker/msg/msg_tlv.py (hunk 3 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py:71]: Magic number(technically MagicWord, but it's not a word) is based on the TI term. It's the fixed value used to identify the start of the TLV header. Keeping the nomenclature consistent with the source code.

File: worker/msg/msg_tlv.py (hunk 4 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py]: Cleaned up

File: worker/msg/msg_tlv.py (hunk 5 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py]: Cleaned up

File: worker/msg/msg_tlv.py (hunk 8 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py]: Cleaned up

File: worker/msg/msg_tlv.py (hunk 9 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py]: Cleaned up

File: worker/msg/msg_tlv.py (hunk 10 above)
Comment (jkey-Ainstein) [reply worker/msg/msg_tlv.py]: Cleaned up

File: worker/worker_wifi.py (hunk 11 above)
Comment (jkey-Ainstein) [reply worker/worker_wifi.py]: Corrected

File: worker/worker_wifi.py (hunk 12 above)
Comment (jkey-Ainstein) [reply worker/worker_wifi.py]: Corrected in a couple places. Good find.

File: worker/msg/msg_tlv.py (hunk 3 above)
Comment (AndyBorders) [reply worker/msg/msg_tlv.py:71]: OK does not mean I like it.  It is really a header/message start.  There is nothing magic about it.

**Pull request merged**